flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageProducerClient.java

```diff
@@ -119,7 +149,60 @@ private void writeAccumulatedBuffers(
     private void writeAccumulatedBuffer(
             TieredStorageSubpartitionId subpartitionId, Buffer accumulatedBuffer)
             throws IOException {
-        // TODO, Try to write the accumulated buffer to the appropriate tier. After the tier is
-        // decided, then write the accumulated buffer to the tier.
+        updateStatistics(accumulatedBuffer);
+        Buffer compressedBuffer = compressBufferIfPossible(accumulatedBuffer);
+
+        if (currentSubpartitionTierAgent[subpartitionId.getSubpartitionId()] == null) {
+            chooseStorageTierToStartSegment(subpartitionId);
+        }
+
+        boolean isSuccess =
+                currentSubpartitionTierAgent[subpartitionId.getSubpartitionId()].write(
+                        subpartitionId, compressedBuffer);
+        if (!isSuccess) {
+            chooseStorageTierToStartSegment(subpartitionId);
+            isSuccess =
+                    currentSubpartitionTierAgent[subpartitionId.getSubpartitionId()].write(
+                            subpartitionId, compressedBuffer);
+            checkState(isSuccess, "Failed to write the first buffer to the new segment");
+        }
+    }
+
+    private void chooseStorageTierToStartSegment(TieredStorageSubpartitionId subpartitionId)
+            throws IOException {
+        int subpartitionIndex = subpartitionId.getSubpartitionId();
+        int segmentIndex = currentSubpartitionSegmentId[subpartitionIndex];
+        int nextSegmentIndex = segmentIndex + 1;
+
+        for (TierProducerAgent tierProducerAgent : tierProducerAgents) {
+            if (tierProducerAgent.tryStartNewSegment(subpartitionId, nextSegmentIndex)) {
+                // Update the segment index and the chosen storage tier for the subpartition.
+                currentSubpartitionSegmentId[subpartitionIndex] = nextSegmentIndex;
+                currentSubpartitionTierAgent[subpartitionIndex] = tierProducerAgent;
+                return;
+            }
+        }
+        throw new IOException("Failed to choose a storage tier to start a new segment.");
+    }
+
+    private Buffer compressBufferIfPossible(Buffer buffer) {
+        if (!canBeCompressed(buffer)) {
+            return buffer;
+        }
+
+        return checkNotNull(bufferCompressor).compressToOriginalBuffer(buffer);
+    }
+
+    /**
+     * Whether the buffer can be compressed or not. Note that event is not compressed because it is
+     * usually small and the size can become even larger after compression.
+     */
+    private boolean canBeCompressed(Buffer buffer) {
+        return bufferCompressor != null && buffer.isBuffer() && buffer.readableBytes() > 0;
+    }
+
+    private void updateStatistics(Buffer buffer) {
+        checkNotNull(outputMetrics).getNumBuffersOut().inc();
+        checkNotNull(outputMetrics).getNumBytesOut().inc(buffer.readableBytes());
```

Comment:
This indicates Flink's metric mechanism invading tiered storage. Alternatively, we can apply a listener pattern that tiered result partition subscribes to the tiered storage statistic updates.

Reply:
Fixed by introducing a `Consumer<TieredStorageProducerMetricStatistics> metricStatisticsUpdater` in the producer client. 

The `TieredResultPartition` will listen to the metric updater.